torch/distributions/transforms.py

```diff
@@ -0,0 +1,231 @@
+import torch
+from torch.distributions import constraints
+from torch.distributions.utils import broadcast_all
+from torch.nn.functional import sigmoid
+
+__all__ = [
+    'AbsTransform',
+    'AffineTransform',
+    'ExpTransform',
+    'InverseTransform',
+    'LogprobTransform',
+    'SigmoidTransform',
+    'StickBreakingTransform',
+    'Transform',
+]
+
+
+class Transform(object):
+    """
+    Abstract class for transformations with computable inverse log
+    det jacobians. They are primarily used in
+    :class:`torch.distributions.TransformedDistribution`.
+
+    Transforms are intended to be short-lived objects. They memoize the forward
+    and inverse computations to avoid work; therefore :meth:`inverse` is
+    nearly free after calling :meth:`forward`. To clear the memoization cache,
+    delete the object and create a new object.
+
+    Derived classes should implement one or both of :meth:`_forward` or
+    :meth:`_inverse` and should implement :meth:`log_abs_det_jacobian`.
+    Derived classes may store intermediate results in the `._cache` dict.
+    """
+
+    def __init__(self):
+        self._cache = {}

```

Comment:
I'm open to ideas here. This seemed like the least weird implementation I could find.

The purposes are (1) to speed up inverses, (2) to compute inverses in cases where they have not been implemented, and (3) cache intermediate computations for computing `log_abs_det_jacobian()` to avoid gradients being ping-ponged back and forth through `.forward()` and `.inverse()` (e.g. if `.log_deg_jacobian_inverse(x, y)` depends on x, we want to avoid `x` being defined in terms of `dist(inverse(dist.forward(y))`. The Pyro implementation automatically ejects an item from cache the first time it is read, but that behavior seemed to make it too easy to introduce a leak.

We could implement cache size of 1, but that could lead to silent degradation of performance and gradient stability. Alican originally had something like a `.memoize` flag to configure this, but I removed it at some point to simplify the interface.

Reply:
Just to clarify: sometimes `.inverse()` is not implemented, and sometimes it is 100x more expensive to compute than to memoize (e.g. `gesv` vs `mm`).

Reply:
I think this makes sense, but I'm worried that this will lead to space leaks - users might not expect us to hold on to memory indefinitely.

Reply:
Ok after design discussion with @martinjankowiak I've changed this to:
- Do no caching by default in `Transform`
- Split out `CachedTransform` that implements a LRU(1) cache
- Only a couple of the more complex transforms inherit from `CachedTransform`, but in the future many more will, e.g. `NormalizingFlow`, `NADE`, ...